{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "from tensorflow.keras.datasets import mnist\n",
    "from tensorflow.keras import layers, models\n",
    "from tensorflow.keras.utils import to_categorical\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import random\n",
    "import time\n",
    "from multiprocessing import Pool\n",
    "from numba import jit, njit, prange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "def softmax(x):\n",
    "    z = x - np.max(x)   # overflow protection (softmax(x) = softmax(x - const))\n",
    "    return np.exp(z) / np.sum(np.exp(z))\n",
    "\n",
    "activation_functions = {\n",
    "    'tanh': np.tanh,\n",
    "    'relu': (lambda x: np.maximum(0, x)),\n",
    "    'sigmoid': (lambda x: 1 / (1 + np.exp(-x))),\n",
    "    'linear': (lambda x: x),\n",
    "    'softmax': softmax\n",
    "}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Loading Data**"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: total: 359 ms\n",
      "Wall time: 357 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "t0 = time.time()\n",
    "(X_train, y_train), (X_test, y_test) = mnist.load_data()\n",
    "X_train = X_train.reshape(-1, np.shape(X_train)[1] * np.shape(X_train)[1]).astype(np.float32) / 255.0\n",
    "X_test = X_test.reshape(-1, np.shape(X_test)[1] * np.shape(X_test)[1]).astype(np.float32) / 255.0\n",
    "y_train = to_categorical(y_train)   # one-hot encoding\n",
    "y_test = to_categorical(y_test)  # one-hot encoding"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# ---"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "# TODO: for debugging only\n",
    "connectivity_matrix1 = np.random.randint(0, 1, size=(32, 784)).astype(np.float32)\n",
    "connectivity_matrix2 = np.random.randint(0, 1, size=(10, 32)).astype(np.float32)\n",
    "bias = np.asarray(1).astype(np.float32)\n",
    "weight = np.asarray(1).astype(np.float32)\n",
    "activation_function1 = activation_functions['sigmoid']\n",
    "activation_function2 = activation_functions['softmax']"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "outputs": [],
   "source": [
    "@jit(parallel=True)\n",
    "def test():\n",
    "    accuracy = 0\n",
    "    for x, y in zip(X_test, y_test):\n",
    "        y_pred = np.argmax(activation_function2(bias + weight * connectivity_matrix2 @ activation_function1(bias + weight * connectivity_matrix1 @ x)))\n",
    "\n",
    "        y_true = np.argmax(y)\n",
    "        if y_pred == y_true:\n",
    "            accuracy += 1"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12.8 s ± 96.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "for i in range(10):\n",
    "    test()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "@jit(parallel=True)\n",
    "def test_ten():\n",
    "    for i in prange(10):\n",
    "        accuracy = 0\n",
    "        for x, y in zip(X_test, y_test):\n",
    "            '''hidden = activation_function1(bias + weight * connectivity_matrix1 @ x)\n",
    "            out = activation_function2(bias + weight * connectivity_matrix2 @ hidden)'''\n",
    "            hidden = bias + weight * connectivity_matrix1 @ x\n",
    "            out = bias + weight * connectivity_matrix2 @ hidden\n",
    "\n",
    "            y_pred = np.argmax(out)\n",
    "            y_true = np.argmax(y)\n",
    "            if y_pred == y_true:\n",
    "                accuracy += 1"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py:8: NumbaPerformanceWarning: \u001B[1m\u001B[1m'@' is faster on contiguous arrays, called on (array(float32, 2d, C), readonly array(float32, 1d, A))\u001B[0m\u001B[0m\n",
      "  hidden = bias + weight * connectivity_matrix1 @ x\n",
      "C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py:8: NumbaPerformanceWarning: \u001B[1m\u001B[1m'@' is faster on contiguous arrays, called on (array(float32, 2d, C), readonly array(float32, 1d, A))\u001B[0m\u001B[0m\n",
      "  hidden = bias + weight * connectivity_matrix1 @ x\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typing\\npydecl.py:965: NumbaPerformanceWarning: \u001B[1m'@' is faster on contiguous arrays, called on (array(float32, 2d, C), readonly array(float32, 1d, A))\u001B[0m\n",
      "  warnings.warn(NumbaPerformanceWarning(msg))\n",
      "<string>:1: NumbaWarning: \u001B[1m\n",
      "Compilation is falling back to object mode WITH looplifting enabled because Function __numba_parfor_gufunc_0x1bccbfa4070 failed at nopython mode lowering due to: \u001B[1m\u001B[1mscipy 0.16+ is required for linear algebra\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 8:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "\u001B[0m\u001B[1mDuring: lowering \"$46binary_matrix_multiply.11 = $42binary_multiply.9 <built-in function matmul> $x.110\" at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (8)\u001B[0m\u001B[0m\n",
      "C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py:1: NumbaWarning: \u001B[1m\n",
      "Compilation is falling back to object mode WITH looplifting enabled because Function test_ten failed at nopython mode lowering due to: \u001B[1mFailed in object mode pipeline (step: remove phis nodes)\n",
      "\u001B[1mIllegal IR, del found at: del $const6.2.205\n",
      "\u001B[1m\n",
      "File \"<string>\", line 2:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "\u001B[0m\u001B[1mDuring: lowering \"id=6[LoopNest(index_variable = parfor_index.95, range = (0, $const4.1, 1))]{226: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (14)>, 12: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (3)>, 311: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (5)>, 313: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (14)>, 28: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (5)>, 30: <ir.Block at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (5)>}Var(parfor_index.95, 1860798773.py:3)\" at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (3)\u001B[0m\u001B[0m\n",
      "  @jit(parallel=True)\n",
      "C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py:1: NumbaWarning: \u001B[1m\n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"test_ten\" failed type inference due to: \u001B[1m\u001B[1mCannot determine Numba type of <class 'numba.core.dispatcher.LiftedLoop'>\u001B[0m\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\u001B[0m\n",
      "  @jit(parallel=True)\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\object_mode_passes.py:151: NumbaWarning: \u001B[1mFunction \"test_ten\" was compiled in object mode without forceobj=True, but has lifted loops.\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\object_mode_passes.py:161: NumbaDeprecationWarning: \u001B[1m\n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n",
      "C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py:1: NumbaWarning: \u001B[1m\n",
      "Compilation is falling back to object mode WITHOUT looplifting enabled because Function \"test_ten\" failed type inference due to: \u001B[1m\u001B[1mnon-precise type pyobject\u001B[0m\n",
      "\u001B[0m\u001B[1mDuring: typing of argument at C:\\Users\\jjung\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py (3)\u001B[0m\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  @jit(parallel=True)\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\object_mode_passes.py:151: NumbaWarning: \u001B[1mFunction \"test_ten\" was compiled in object mode without forceobj=True.\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaWarning(warn_msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\object_mode_passes.py:161: NumbaDeprecationWarning: \u001B[1m\n",
      "Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.\n",
      "\n",
      "For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit\n",
      "\u001B[1m\n",
      "File \"..\\..\\..\\..\\..\\..\\AppData\\Local\\Temp\\ipykernel_17552\\1860798773.py\", line 3:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaDeprecationWarning(msg,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.42 s ± 194 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "test_ten()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# ----------------------------------------- DEBUG"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  C_CONTIGUOUS : True\n",
      "  F_CONTIGUOUS : False\n",
      "  OWNDATA : True\n",
      "  WRITEABLE : True\n",
      "  ALIGNED : True\n",
      "  WRITEBACKIFCOPY : False\n",
      "  UPDATEIFCOPY : False\n",
      " \n",
      "   C_CONTIGUOUS : True\n",
      "  F_CONTIGUOUS : False\n",
      "  OWNDATA : True\n",
      "  WRITEABLE : True\n",
      "  ALIGNED : True\n",
      "  WRITEBACKIFCOPY : False\n",
      "  UPDATEIFCOPY : False\n",
      "\n",
      "(32, 784) (10000, 784)\n"
     ]
    }
   ],
   "source": [
    "A = np.random.randint(0, 1, size=(32, 784)).astype(np.float32)\n",
    "\n",
    "A = np.ascontiguousarray(A)\n",
    "X = np.ascontiguousarray(X_test)\n",
    "print(\n",
    "    A.flags, '\\n',\n",
    "    X.flags\n",
    ")\n",
    "print(A.shape, X.shape)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40.7 ms ± 1.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n",
      "C:\\Users\\jjung\\anaconda3\\envs\\TF_CPU\\lib\\site-packages\\numba\\core\\typed_passes.py:329: NumbaPerformanceWarning: \u001B[1m\n",
      "The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.\n",
      "\n",
      "To find out why, try turning on parallel diagnostics, see https://numba.readthedocs.io/en/stable/user/parallel.html#diagnostics for help.\n",
      "\u001B[1m\n",
      "File \"<magic-timeit>\", line 1:\u001B[0m\n",
      "\u001B[1m<source missing, REPL/exec in use?>\u001B[0m\n",
      "\u001B[0m\n",
      "  warnings.warn(errors.NumbaPerformanceWarning(msg,\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "@njit(parallel=True)\n",
    "def foo():\n",
    "    s=0\n",
    "    for j in prange(X.shape[0]):\n",
    "        x = X[j]\n",
    "        A @ x\n",
    "foo()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 µs ± 3.58 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "#@njit\n",
    "def foobar():\n",
    "    A @ X[0]\n",
    "\n",
    "foobar()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# ----------------------------------------- DEBUG"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12.6 s ± 30 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "for i in range(10):\n",
    "    accuracy = 0\n",
    "    for x, y in zip(X_test, y_test):\n",
    "        y_pred = np.argmax(activation_function2(bias + weight * connectivity_matrix2 @ activation_function1(bias + weight * connectivity_matrix1 @ x)))\n",
    "\n",
    "        y_true = np.argmax(y)\n",
    "        if y_pred == y_true:\n",
    "            accuracy += 1"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "%%time\n",
    "pool = Pool(6)\n",
    "pool.starmap(test, [() for i in range(6)])"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": true
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# ---"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Network Definition**"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "class Layer():\n",
    "    def __init__(self, weight, connectivity_matrix, bias, activation_function,\n",
    "                 innovation, enabled, output_layer=False):\n",
    "        self.weight = weight\n",
    "        self.connectivity_matrix = connectivity_matrix\n",
    "        self.bias = bias\n",
    "        if output_layer:\n",
    "            assert(activation_function == 'softmax'), 'output layer must use softmax'\n",
    "        self.activation_function = activation_function\n",
    "        self.innovation = innovation\n",
    "        self.enabled = enabled\n",
    "        self.output_layer = output_layer    # final 10 neuron layer (activation fixed to softmax)\n",
    "\n",
    "        # TODO: tune\n",
    "        self.mutate_rate_weight = 0.1\n",
    "        self.mutate_rate_connections = 0.3\n",
    "        self.mutate_rate_bias = 0.1\n",
    "        self.mutate_rate_activation_function = 0.1\n",
    "\n",
    "    def forward(self, input):\n",
    "        return activation_functions[self.activation_function](\n",
    "            self.weight * self.connectivity_matrix @ input + self.bias\n",
    "        )\n",
    "\n",
    "    def mutate(self):\n",
    "        # TODO: correct idea?\n",
    "        # TODO: quite arbitrary, inefficient and do negative numbers work?\n",
    "\n",
    "        # weight\n",
    "        # TODO: change this mess\n",
    "        weight_bin = bin(self.weight)[2:].zfill(7)   # 7 bit bitstring\n",
    "        temp = ''\n",
    "        for i in range(len(weight_bin)):\n",
    "            if random.uniform(0, 1) < self.mutate_rate_weight:\n",
    "                temp2 = '0' if weight_bin[i] == '1' else '1'\n",
    "                temp += temp2\n",
    "            else:\n",
    "                temp += weight_bin[i]\n",
    "        self.weight = int(temp, 2)\n",
    "\n",
    "        # connectivity matrix\n",
    "        # TODO: row, col correct?\n",
    "        for row in range(np.shape(self.connectivity_matrix)[0]):\n",
    "            for col in range(np.shape(self.connectivity_matrix)[1]):\n",
    "                if random.uniform(0, 1) < self.mutate_rate_connections:\n",
    "                    self.connectivity_matrix[row][col] = 1 - self.connectivity_matrix[row][col]\n",
    "\n",
    "        # bias\n",
    "        # TODO: change this mess\n",
    "        bias_bin = bin(self.bias)[2:].zfill(7)  # 7 bit bitstring\n",
    "        temp = ''\n",
    "        for i in range(len(bias_bin)):\n",
    "            if random.uniform(0, 1) < self.mutate_rate_bias:\n",
    "                temp2 = '0' if bias_bin[i] == '1' else '1'\n",
    "                temp += temp2\n",
    "            else:\n",
    "                temp += bias_bin[i]\n",
    "        self.bias = int(temp, 2)\n",
    "\n",
    "        # activation function\n",
    "        if not self.output_layer:\n",
    "            if random.uniform(0, 1) < self.mutate_rate_activation_function:\n",
    "                self.activation_function = random.choice(list(activation_functions.keys()))"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "class Network():\n",
    "    def __init__(self, layers):\n",
    "        self.layers = layers    # structured (feedforward) array of layers\n",
    "\n",
    "    def forward(self, input):\n",
    "        temp = input\n",
    "        for layer in self.layers:\n",
    "            temp = layer.forward(temp)\n",
    "        return temp\n",
    "\n",
    "    def evaluate(self):  # TODO: slow? (on cpu ~0.55s per call)\n",
    "        accuracy = 0\n",
    "        for x, y in zip(X_test, y_test):\n",
    "            y_pred = np.argmax(self.forward(x.reshape(-1)))  # class with highest value\n",
    "            y_true = np.argmax(y)\n",
    "            if y_pred == y_true:\n",
    "                accuracy += 1\n",
    "        return accuracy / np.shape(X_test)[0]\n",
    "\n",
    "    def mutate(self):\n",
    "        for layer in self.layers:\n",
    "            layer.mutate()\n",
    "        # TODO: change network mutation (new layers)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [],
   "source": [
    "class Population():\n",
    "    def __init__(self, size=10, n_survivors=5):\n",
    "        self.generation = 0\n",
    "        self.size = size\n",
    "        self.n_survivors = n_survivors  # TODO: for now must be odd number (1 elite + odd breeders)\n",
    "        self.elite = None\n",
    "\n",
    "        # initialization # TODO: better (quite random now - e.g. 32 hidden neurons)\n",
    "        self.organisms = []\n",
    "        for i in range(size):\n",
    "            # one hidden layer (32 neurons), one output layer (10 neurons - 1 resp. to each class)\n",
    "            self.organisms.append(Network([\n",
    "                Layer(\n",
    "                    weight=1,\n",
    "                    connectivity_matrix=np.round(np.random.rand(32, 784)),\n",
    "                    bias=0,\n",
    "                    activation_function='sigmoid',\n",
    "                    innovation=0,\n",
    "                    enabled=True,\n",
    "                    output_layer=False\n",
    "                ),\n",
    "                Layer(\n",
    "                    weight=1,\n",
    "                    connectivity_matrix=np.round(np.random.rand(32, 32)),\n",
    "                    bias=0,\n",
    "                    activation_function='sigmoid',\n",
    "                    innovation=0,\n",
    "                    enabled=True,\n",
    "                    output_layer=False\n",
    "                ),\n",
    "                Layer(\n",
    "                    weight=1,\n",
    "                    connectivity_matrix=np.round(np.random.rand(32, 32)),\n",
    "                    bias=0,\n",
    "                    activation_function='sigmoid',\n",
    "                    innovation=0,\n",
    "                    enabled=True,\n",
    "                    output_layer=False\n",
    "                ),\n",
    "                Layer(\n",
    "                    weight=1,\n",
    "                    connectivity_matrix=np.round(np.random.rand(32, 32)),\n",
    "                    bias=0,\n",
    "                    activation_function='sigmoid',\n",
    "                    innovation=0,\n",
    "                    enabled=True,\n",
    "                    output_layer=False\n",
    "                ),\n",
    "                Layer(\n",
    "                    weight=1,\n",
    "                    connectivity_matrix=np.round(np.random.rand(10, 32)),\n",
    "                    bias=0,\n",
    "                    activation_function='softmax',\n",
    "                    innovation=0,\n",
    "                    enabled=True,\n",
    "                    output_layer=True\n",
    "                )\n",
    "            ]))\n",
    "\n",
    "        self.history = [(max(self.organism_fitness()), self.average_fitness())]   # fitness of population over all generations\n",
    "\n",
    "    def organism_fitness(self):\n",
    "        # TODO: speciation\n",
    "        return [organism.evaluate() for organism in self.organisms]\n",
    "\n",
    "    def average_fitness(self):\n",
    "        # TODO: speciation\n",
    "        organism_fitness = self.organism_fitness()\n",
    "        return sum(organism_fitness) / len(organism_fitness)\n",
    "\n",
    "    def max_fitness(self):\n",
    "        # TODO: speciation\n",
    "        return max(self.organism_fitness())\n",
    "\n",
    "    def selection(self):\n",
    "        organism_fitness = self.organism_fitness()\n",
    "\n",
    "        # elitism (n=1)\n",
    "        elite_index = np.argmax(organism_fitness)\n",
    "        self.elite = self.organisms.pop(elite_index)\n",
    "        organism_fitness.pop(elite_index)\n",
    "\n",
    "        probabilities = [fitness / sum(organism_fitness) for fitness in organism_fitness]  # normalized\n",
    "        survivors = np.random.choice(self.organisms,\n",
    "                                     size=self.n_survivors - 1,\n",
    "                                     p=probabilities,\n",
    "                                     replace=False)    # TODO: works without replacement and p?\n",
    "        return [survivor for survivor in survivors]\n",
    "\n",
    "    def crossover(self, parents):\n",
    "        # TODO: for different type of networks\n",
    "        # TODO: correct?\n",
    "        children = []\n",
    "        while len(children) < (self.size - 1):\n",
    "            [father, mother] = random.sample(parents + [self.elite], k=2)  # sample without replacement\n",
    "            layers = []\n",
    "\n",
    "            # TODO: for now assume same no of layers\n",
    "            for father_gene, mother_gene in zip(father.layers, mother.layers):\n",
    "                # full gene crossover # TODO: correct?\n",
    "                child_weight = father_gene.weight if (random.uniform(0, 1) < 0.5) else mother_gene.weight\n",
    "                child_bias = father_gene.bias if (random.uniform(0, 1) < 0.5) else mother_gene.bias\n",
    "                child_activation_function = father_gene.activation_function if \\\n",
    "                    (random.uniform(0, 1) < 0.5) else mother_gene.activation_function\n",
    "\n",
    "                # uniform (bit-wise) crossover # TODO: correct?\n",
    "                child_connectivity_matrix = np.zeros(np.shape(father_gene.connectivity_matrix))\n",
    "                for row in range(np.shape(child_connectivity_matrix)[0]):\n",
    "                    for col in range(np.shape(child_connectivity_matrix)[1]):\n",
    "                        child_connectivity_matrix[row][col] = father_gene.connectivity_matrix[row][col] \\\n",
    "                            if (random.uniform(0, 1) < 0.5) else mother_gene.connectivity_matrix[row][col]\n",
    "\n",
    "                layers.append(Layer(\n",
    "                    weight=child_weight,\n",
    "                    connectivity_matrix=child_connectivity_matrix,\n",
    "                    bias=child_bias,\n",
    "                    activation_function=child_activation_function,\n",
    "                    innovation=0,\n",
    "                    enabled=True\n",
    "                ))\n",
    "\n",
    "            children.append(Network(layers))\n",
    "\n",
    "        return children\n",
    "\n",
    "    def mutate(self, organisms):\n",
    "        for organism in organisms:\n",
    "            organism.mutate()\n",
    "\n",
    "    def breed(self):\n",
    "        t = time.time()\n",
    "        parents = self.selection()\n",
    "        print('selection time: {}s'.format(round(time.time() - t, 3)))\n",
    "        t = time.time()\n",
    "        children = self.crossover(parents)\n",
    "        print('crossover time: {}s'.format(round(time.time() - t, 3)))\n",
    "        t = time.time()\n",
    "        '''print('pre mutation', [child.evaluate() for child in children], 'elite', self.elite.evaluate())\n",
    "        self.mutate(children)\n",
    "        print('after mutation', [child.evaluate() for child in children], 'elite', self.elite.evaluate())\n",
    "        weights = '['\n",
    "        for organims in (children + [self.elite]):\n",
    "            weights += str(organims.layers[0].weight) + ', '\n",
    "        print(weights + ']')'''\n",
    "        self.organisms = children + [self.elite]\n",
    "        self.generation += 1\n",
    "        self.history.append((self.max_fitness(), self.average_fitness()))\n",
    "\n",
    "    def plot(self):\n",
    "        plt.figure()\n",
    "        plt.plot(np.arange(self.generation + 1), [score[0] for score in self.history],\n",
    "                 label='max fitness')\n",
    "        plt.plot(np.arange(self.generation + 1), [score[1] for score in self.history],\n",
    "                 label='avg fitness', alpha=0.6)\n",
    "        plt.title('Population fitness')\n",
    "        plt.xlabel('Generations')\n",
    "        plt.ylabel('Fitness score (accuracy)')\n",
    "        plt.legend()\n",
    "        plt.show()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**Training**"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [],
   "source": [
    "# initialization\n",
    "GENERATIONS = 50\n",
    "POPULATION_SIZE = 10\n",
    "SURVIVORS = 5\n",
    "population = Population(size=POPULATION_SIZE, n_survivors=SURVIVORS)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting training\n",
      "Gen 0 : [0.0974, 0.1135, 0.1032, 0.1009, 0.1135, 0.1009, 0.1032, 0.1032, 0.0892, 0.1032] - max: 0.1135 (31.77s)\n",
      "selection time: 15.78s\n",
      "crossover time: 0.224s\n",
      "evaluation time: 30.365s\n",
      "Gen 1 : [0.1009, 0.098, 0.1032, 0.1032, 0.1032, 0.098, 0.1135, 0.1032, 0.1135, 0.1135] - max: 0.1135 (78.85s)\n",
      "selection time: 16.16s\n",
      "crossover time: 0.214s\n",
      "evaluation time: 32.812s\n",
      "Gen 2 : [0.098, 0.1135, 0.1135, 0.0892, 0.1135, 0.0982, 0.1028, 0.1009, 0.0958, 0.1135] - max: 0.1135 (81.26s)\n",
      "selection time: 15.965s\n",
      "crossover time: 0.219s\n",
      "evaluation time: 32.208s\n",
      "Gen 3 : [0.1135, 0.1028, 0.1032, 0.1028, 0.1135, 0.1032, 0.1135, 0.1135, 0.1135, 0.1135] - max: 0.1135 (79.9s)\n",
      "selection time: 15.745s\n",
      "crossover time: 0.22s\n",
      "evaluation time: 32.431s\n",
      "Gen 4 : [0.1028, 0.1032, 0.1028, 0.1135, 0.1032, 0.1032, 0.1135, 0.1028, 0.1028, 0.1135] - max: 0.1135 (79.89s)\n",
      "selection time: 15.092s\n",
      "crossover time: 0.284s\n",
      "evaluation time: 30.083s\n",
      "Gen 5 : [0.1028, 0.1135, 0.1028, 0.1028, 0.1032, 0.1135, 0.1032, 0.1135, 0.1135, 0.1135] - max: 0.1135 (75.9s)\n",
      "selection time: 14.924s\n",
      "crossover time: 0.211s\n",
      "evaluation time: 30.08s\n",
      "Gen 6 : [0.1028, 0.1135, 0.1135, 0.1028, 0.1135, 0.1028, 0.1028, 0.1032, 0.1032, 0.1135] - max: 0.1135 (75.36s)\n",
      "selection time: 14.959s\n",
      "crossover time: 0.22s\n",
      "evaluation time: 29.769s\n",
      "Gen 7 : [0.1032, 0.1135, 0.1135, 0.1028, 0.1032, 0.1032, 0.1135, 0.1032, 0.1028, 0.1135] - max: 0.1135 (74.96s)\n",
      "selection time: 14.813s\n",
      "crossover time: 0.215s\n",
      "evaluation time: 29.842s\n",
      "Gen 8 : [0.1028, 0.1135, 0.1135, 0.1135, 0.1032, 0.1135, 0.1135, 0.1028, 0.1135, 0.1135] - max: 0.1135 (75.07s)\n",
      "selection time: 14.892s\n",
      "crossover time: 0.227s\n",
      "evaluation time: 29.453s\n",
      "Gen 9 : [0.1135, 0.1028, 0.1028, 0.1028, 0.1135, 0.1032, 0.1135, 0.1032, 0.1032, 0.1135] - max: 0.1135 (73.78s)\n",
      "selection time: 14.895s\n",
      "crossover time: 0.221s\n",
      "evaluation time: 31.111s\n",
      "Gen 10 : [0.1135, 0.1135, 0.1032, 0.1032, 0.1028, 0.1135, 0.1028, 0.1028, 0.1135, 0.1135] - max: 0.1135 (77.45s)\n",
      "selection time: 15.495s\n",
      "crossover time: 0.22s\n",
      "evaluation time: 34.243s\n",
      "Gen 11 : [0.1135, 0.1028, 0.1135, 0.1135, 0.1135, 0.1135, 0.1032, 0.1032, 0.1032, 0.1135] - max: 0.1135 (82.69s)\n",
      "selection time: 17.634s\n",
      "crossover time: 0.22s\n",
      "evaluation time: 34.708s\n",
      "Gen 12 : [0.1135, 0.1135, 0.1028, 0.1032, 0.1032, 0.1135, 0.1135, 0.1028, 0.1135, 0.1135] - max: 0.1135 (87.31s)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Input \u001B[1;32mIn [8]\u001B[0m, in \u001B[0;36m<cell line: 13>\u001B[1;34m()\u001B[0m\n\u001B[0;32m     13\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m generation \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mrange\u001B[39m(\u001B[38;5;241m1\u001B[39m, GENERATIONS):\n\u001B[0;32m     14\u001B[0m     \u001B[38;5;66;03m# breed new population\u001B[39;00m\n\u001B[0;32m     15\u001B[0m     t1 \u001B[38;5;241m=\u001B[39m time\u001B[38;5;241m.\u001B[39mtime()\n\u001B[1;32m---> 16\u001B[0m     \u001B[43mpopulation\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mbreed\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     18\u001B[0m     \u001B[38;5;66;03m# evaluate new population\u001B[39;00m\n\u001B[0;32m     19\u001B[0m     ta\u001B[38;5;241m=\u001B[39mtime\u001B[38;5;241m.\u001B[39mtime()\n",
      "Input \u001B[1;32mIn [6]\u001B[0m, in \u001B[0;36mPopulation.breed\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m    130\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mbreed\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[0;32m    131\u001B[0m     t \u001B[38;5;241m=\u001B[39m time\u001B[38;5;241m.\u001B[39mtime()\n\u001B[1;32m--> 132\u001B[0m     parents \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mselection\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m    133\u001B[0m     \u001B[38;5;28mprint\u001B[39m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mselection time: \u001B[39m\u001B[38;5;132;01m{}\u001B[39;00m\u001B[38;5;124ms\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mformat(\u001B[38;5;28mround\u001B[39m(time\u001B[38;5;241m.\u001B[39mtime() \u001B[38;5;241m-\u001B[39m t, \u001B[38;5;241m3\u001B[39m)))\n\u001B[0;32m    134\u001B[0m     t \u001B[38;5;241m=\u001B[39m time\u001B[38;5;241m.\u001B[39mtime()\n",
      "Input \u001B[1;32mIn [6]\u001B[0m, in \u001B[0;36mPopulation.selection\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m     75\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mselection\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[1;32m---> 76\u001B[0m     organism_fitness \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43morganism_fitness\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     78\u001B[0m     \u001B[38;5;66;03m# elitism (n=1)\u001B[39;00m\n\u001B[0;32m     79\u001B[0m     elite_index \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39margmax(organism_fitness)\n",
      "Input \u001B[1;32mIn [6]\u001B[0m, in \u001B[0;36mPopulation.organism_fitness\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m     62\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21morganism_fitness\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[0;32m     63\u001B[0m     \u001B[38;5;66;03m# TODO: speciation\u001B[39;00m\n\u001B[1;32m---> 64\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m [organism\u001B[38;5;241m.\u001B[39mevaluate() \u001B[38;5;28;01mfor\u001B[39;00m organism \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39morganisms]\n",
      "Input \u001B[1;32mIn [6]\u001B[0m, in \u001B[0;36m<listcomp>\u001B[1;34m(.0)\u001B[0m\n\u001B[0;32m     62\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21morganism_fitness\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[0;32m     63\u001B[0m     \u001B[38;5;66;03m# TODO: speciation\u001B[39;00m\n\u001B[1;32m---> 64\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m [\u001B[43morganism\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mevaluate\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m \u001B[38;5;28;01mfor\u001B[39;00m organism \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39morganisms]\n",
      "Input \u001B[1;32mIn [5]\u001B[0m, in \u001B[0;36mNetwork.evaluate\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m     12\u001B[0m accuracy \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m0\u001B[39m\n\u001B[0;32m     13\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m x, y \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mzip\u001B[39m(X_test, y_test):\n\u001B[1;32m---> 14\u001B[0m     y_pred \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39margmax(\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mforward\u001B[49m\u001B[43m(\u001B[49m\u001B[43mx\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mreshape\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;241;43m-\u001B[39;49m\u001B[38;5;241;43m1\u001B[39;49m\u001B[43m)\u001B[49m\u001B[43m)\u001B[49m)  \u001B[38;5;66;03m# class with highest value\u001B[39;00m\n\u001B[0;32m     15\u001B[0m     y_true \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39margmax(y)\n\u001B[0;32m     16\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m y_pred \u001B[38;5;241m==\u001B[39m y_true:\n",
      "Input \u001B[1;32mIn [5]\u001B[0m, in \u001B[0;36mNetwork.forward\u001B[1;34m(self, input)\u001B[0m\n\u001B[0;32m      6\u001B[0m temp \u001B[38;5;241m=\u001B[39m \u001B[38;5;28minput\u001B[39m\n\u001B[0;32m      7\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m layer \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mlayers:\n\u001B[1;32m----> 8\u001B[0m     temp \u001B[38;5;241m=\u001B[39m \u001B[43mlayer\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mforward\u001B[49m\u001B[43m(\u001B[49m\u001B[43mtemp\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m      9\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m temp\n",
      "Input \u001B[1;32mIn [4]\u001B[0m, in \u001B[0;36mLayer.forward\u001B[1;34m(self, input)\u001B[0m\n\u001B[0;32m     20\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mforward\u001B[39m(\u001B[38;5;28mself\u001B[39m, \u001B[38;5;28minput\u001B[39m):\n\u001B[0;32m     21\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m activation_functions[\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mactivation_function](\n\u001B[1;32m---> 22\u001B[0m         \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mweight\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mconnectivity_matrix\u001B[49m \u001B[38;5;241m@\u001B[39m \u001B[38;5;28minput\u001B[39m \u001B[38;5;241m+\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mbias\n\u001B[0;32m     23\u001B[0m     )\n",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": [
    "# initial population\n",
    "print('Starting training')\n",
    "t_training = time.time()\n",
    "population_fitness = population.organism_fitness()\n",
    "max_fitness = population.max_fitness()\n",
    "t2 = time.time()\n",
    "print('Gen', 0, ':',\n",
    "      population_fitness, '- max:',\n",
    "      max_fitness,\n",
    "      '({}s)'.format(round(t2 - t_training, 2)))\n",
    "\n",
    "# future populations\n",
    "for generation in range(1, GENERATIONS):\n",
    "    # breed new population\n",
    "    t1 = time.time()\n",
    "    population.breed()\n",
    "\n",
    "    # evaluate new population\n",
    "    ta=time.time()\n",
    "    population_fitness = population.organism_fitness()\n",
    "    max_fitness = population.max_fitness()\n",
    "    print('evaluation time: {}s'.format(round(time.time() - ta, 3)))\n",
    "    t2 = time.time()\n",
    "\n",
    "    print('Gen', generation, ':',\n",
    "          population_fitness, '- max:',\n",
    "          max_fitness,\n",
    "          '({}s)'.format(round(t2 - t1, 2)))\n",
    "\n",
    "print('Finished training ({})'.format(round(time.time() - t_training, 2)))\n",
    "print('\\nTotal computation time: ({}s)'.format(round(time.time() - t0, 2)))\n",
    "\n",
    "# performance of population\n",
    "population.plot()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}